spyder/plugins/ipythonconsole/widgets/shell.py

```diff
@@ -295,44 +282,61 @@ def is_external_kernel(self):
 
     def setup_spyder_kernel(self):
         """Setup spyder kernel"""
-        self.kernel_handler.open_comm(self.spyder_kernel_comm)
-
-        # For completions
-        self.kernel_client.control_channel.message_received.connect(
-            self._dispatch)
-
-        # Redefine the complete method to work while debugging.
-        self._redefine_complete_for_dbg(self.kernel_client)
+        if not self._init_kernel_setup:
+            self._init_kernel_setup = True
+            # Only do this setup once
+            self.kernel_handler.kernel_comm.sig_exception_occurred.connect(
+                self.sig_exception_occurred)
+            # For completions
+            self.kernel_client.control_channel.message_received.connect(
+                self._dispatch)
+
+            # Redefine the complete method to work while debugging.
+            self._redefine_complete_for_dbg(self.kernel_client)
+
+            for request_id, handler in self.kernel_comm_handlers.items():
+                self.kernel_handler.kernel_comm.register_call_handler(
+                    request_id, handler)
+
+        # Setup to do after restart
+        # Check for fault and send config
+        self.kernel_handler.poll_fault_text()
+
+        # Show possible errors when setting Matplotlib backend
+        self.call_kernel().show_mpl_backend_errors()
+    
+        # Check if the dependecies for special consoles are available.
+        self.call_kernel(
+            callback=self.ipyclient._show_special_console_error
+            ).is_special_kernel_valid()
 
-        # Send configuration
         self.send_spyder_kernel_configuration()
 
-    def pop_execute_queue(self):
-        """Pop one waiting instruction."""
-        if self._execute_queue:
-            self.execute(*self._execute_queue.pop(0))
-
     def send_spyder_kernel_configuration(self):
         """Send kernel configuration to spyder kernel."""
+        # Set current cwd
+        self.set_cwd()
```

Comment:
Why is this necessary here?

Reply:
If the kernel restarts, the kernel loses the cwd, so it must be reset here.